### DISCUSSION TITLE
Get Documents by Batch

### DISCUSSION THREAD

**Discussion by derekperkins** [meilisearch/product#644] [Feedback & Feature Proposal]
Today, there is a [Get One Document](https://www.meilisearch.com/docs/reference/api/documents#get-one-document) endpoint, and a [Delete Documents by Batch](https://www.meilisearch.com/docs/reference/api/documents#delete-documents-by-batch). We have cases where we want to retrieve a large batch of documents by id (~2k), and today we have to loop through and do a call per id, which often causes availability errors. It would be very useful to do a bulk retrieval. I would expect the input to be the same as the batch delete, and the response to be the same as [Get Documents](https://www.meilisearch.com/docs/reference/api/documents#get-documents), so there's nothing really new to architect.

**Comment by gmourier** [comment]
Hello, @derekperkins 👋

What do you think of retrieving them with a filter on the id field? We are currently working on a solution to delete documents by filter, and we have the idea to propose a Get Document by filter method. 

If `id` is defined as a `filterableAttributes`, you will be able to get documents matching the list of ids using the `IN` operator

Could it work for you?

cc @macraig @irevoire

**Reply by derekperkins** [reply]
If the filter can take a list of 10k IDs, I don't particularly care about the implementation details

**Comment by derekperkins** [comment]
Closing the loop here, it looks like it just released in v1.2.0. Thanks for the quick turnaround!

https://github.com/meilisearch/meilisearch/pull/3570
https://github.com/meilisearch/meilisearch/releases/tag/v1.2.0
https://www.meilisearch.com/docs/reference/api/documents#get-documents-with-post

**Reply by tacman** [reply]
I don't think so -- the ID is not a facet (nor should it be), the fixes in 1.2 allowed getting batch documents by filter (aka facets).

Or if I've misunderstood this, can you give  me an example of how to get a batch of documents by id?

Let's say I have an index of 100,000 people and their UID is their social security number.  How do I get results by SSN, without creating a filter on ssn? I think facets are limited to 1000 distint values.

**Reply by derekperkins** [reply]
IIUC, you aren't using faceted search with its limitations. All you have to do is have the id set as a filterable attribute, then use the dedicated endpoint. The docs don't state specifically if there's an upper bound. Are you specifically opposed to adding the id as a filterable attribute?  Seems easy enough to do a quick test on your data.
https://www.meilisearch.com/docs/reference/api/documents#get-documents-with-post

**Reply by tacman** [reply]
A filterable attribute is a facet.  And a facet is not a good approach, as the keys are unique.  So it duplicates the data. 

This seems like a glaring omission.   If I can find the link to add a suggestion, I'll post it here so people can vote on it.  It seems like it wouldn't be that hard, fetching by id already works, so fetching by multiple ids should be easy.

**Reply by derekperkins** [reply]
I'll watch for your suggestion. I get your point that a facet may not be ideal performance / duplication wise, but it's infinitely better than before where we had to fetch one at a time

**Reply by tacman** [reply]
I don't think this issue should be closed.

**Reply by tacman** [reply]
I've added it as a suggestion, but it's not available to vote on yet.

**Reply by derekperkins** [reply]
Ok, reopened

**Reply by tacman** [reply]
I added an item to vote on, but it never showed up in the list, so I assume it needed to be approved.  I can try again.

**Comment by derekperkins** [comment]
I'm reopening this discussion per @tacman wanting to make it available without requiring the id to be a filterable attribute.

**Reply by irevoire** [reply]
Hey, I think we can actually close this discussion as we effectively gave the possibility to get documents by batch via the filters.
And @tacman, I can’t see your suggestion, but the best to help us track it would be to post it here where someone was explicitly asking to get many documents by IDs; https://github.com/meilisearch/product/discussions/99

**Reply by tacman** [reply]
Again, using filters is a hack for smaller collections, but isn't the right solution.  It duplicates the data, and filters aren't set up to handle large quantities of data that way.

For example, when you do a facet distribution query, you're going to get back the entire database with a quanity of "1".  What a waste, I don't want to have the id duplicated in the data or hacking away and "give me the facets except id".

**Reply by irevoire** [reply]
> Again, using filters is a hack for smaller collections, but isn't the right solution. It duplicates the data, and filters aren't set up to handle large quantities of data that way.

That’s only true for the primary key, which wasn’t the main point of this discussion: « Get Documents by Batch ».
I’m not saying your point is not valid; I’m simply saying that if you want a solution to be implemented, the best thing is to help us track and find your comments easily => by posting your message in the right discussion instead of here.

**Reply by tacman** [reply]
I see, so close this not because it's fixed, but because it should be in another thread (and indeed, I've posted there as well and am tracking it.)   This is a big issue for me, the ability to get a batch of documents by identifier.  I'll move to the discussion thread.

**Reply by irevoire** [reply]
Perfect, thanks a lot; it’ll be easier to differentiate and track both issues!
